js/selectors/dashboard-selectors.js

```diff
@@ -44,14 +45,52 @@ export const getStudentProgress = createSelector(
   }
 )
 
-// Returns sorted students.
+// Returns a mapping from student IDs to their total progress across all activities
+export const getStudentTotalProgress = createSelector(
+  [ getStudents, getStudentProgress ],
+  (students, studentProgress) => {
+    return students.map(student => {
+      const id = student.get('id').toString()
+      return studentProgress.get(id).toList().reduce((sum, progress) => sum + progress)
+    })
+  }
+)
+
+// A comparison function to sort students by last and then first name
+const compareStudentsByName = (student1, student2) => {
+  const lastNameCompare = student1.get('lastName').localeCompare(student2.get('lastName'))
+  if (lastNameCompare !== 0) {
+    return lastNameCompare
+  } else {
+    return student1.get('firstName').localeCompare(student2.get('firstName'))
+  }
+}
+
+// Returns sorted students
 export const getSortedStudents = createSelector(
-  [ getStudents, getDashboardSortBy ],
-  (students, sortBy) => {
-    if (sortBy === 'name') {
-      return students.toList().sortBy(student =>
-        (student.get('lastName') + student.get('firstName')).toLowerCase()
-      )
+  [ getStudents, getDashboardSortBy, getStudentTotalProgress ],
```

Comment:
I call the `getStudentTotalProgress` selector here, even though it doesn't always get used when sorting by name. Will this cause unnecessary re-calculations?

Reply:
Yes, but I wouldn't worry about it. It's just one edge case when we sort by name. It's not that expensive to sort students and student total progress won't change that often (even if it's every few seconds).